File: credit.c
#include<stdio.h>
#include<cs50.h>
#include<math.h>


int main()
{
    long n = get_long("Enter Credit Card Number: ");
    int d1, d2, len, m = 0;
    long cc = n;
    long x = n;
    long s = n;
    cc = n;
    while (cc > 0)
    {
        cc = cc / 10;
        m += 1;
    }
    len = m;
    if (len != 13 && len != 15 && len != 16)
    {
        printf("INVALID\n");
        return 0;
    }
    int mod_1, mod_2, sum_1, sum_2;
    sum_1 = sum_2 = 0;
    //Calculate Sum
    do
    {
        //remove last digit
        mod_1 = x % 10;
        x = x / 10;
        sum_1 += mod_1;

        //remove 2nd last digit
        mod_2 = x % 10;
        x = x / 10;
        mod_2 = mod_2 * 2;
        d1 = mod_2 % 10;
        d2 = mod_2 / 10;
        sum_2 = sum_2 + d1 + d2;
    }
    while (x > 0);
    printf("%i\n%i",sum_1,sum_2);
    
    int ttl = sum_1 + sum_2;
    //Check Weather last number is 0 or not
    if (ttl % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }
    //Find out first 2 numbers
    do 
    {
        s = s / 10; 
    }
    while (s > 100);
    //Check wheather Card iiiis Master Card or not
    if ((s / 10 == 5) && (0 < s % 10 && s % 10 < 6))
    {
        printf("MASTER CARD\n");
        return 0;
    }
    //Check wheather Card is Amex or not
    else if (s / 10 == 3 && (s % 10 == 4 || s % 10 == 7))
    {
        printf("AMEX\n");
        return 0;
    } 
    //Check wheather Card is VISA or not
    else if (s / 10 == 4)
    {
        printf("VISA\n");
        return 0;
    } 
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

File: credit.py
import sys


def main():
    number = get_credit()
    validate_number(number)
    
# Program to get credit card Number from user and reprompt in case of invalid input


def get_credit():
    while True:
        number = input("Number: ")
        # check weather number is valid integer and not a sentinal value
        try:
            if len(number) > 0 and int(number):
                break
        except ValueError:
            continue
    return number
    
# Program to calculate checksum and validate for soecific credit card company


def validate_number(number):

    sum1 = sum2 = 0
    # Check length of the number 
    if len(number) not in [13, 14, 15, 16]:
        print("INVALID")
        sys.exit(0)
        
    # check weather number is even or odd to calculate checksum
    if len(number) % 2 == 0:
        for i in range(len(number)):
            # Calculating Sum 1
            if i % 2 != 0:
                sum1 += int(number[i])
            # Calculating Sum 2
            else:
                mul = int(number[i]) * 2
                if mul >= 10:
                    sum2 += mul // 10
                    sum2 += mul % 10
                else:
                    sum2 += mul
                
    else:
        for i in range(len(number)):
            # Calculating Sum 1
            if i % 2 == 0:
                sum1 += int(number[i])
                # Calculating Sum 2
            else:
                mul = int(number[i]) * 2
                if mul >= 10:
                    sum2 += mul // 10
                    sum2 += mul % 10
                else:
                    sum2 += mul
 
    ttl = sum1 + sum2
    if ttl % 10 != 0:
        print("INVALID")
        sys.exit(0)
    
    first = int(number[0])
    second = int(number[1])
    
    # check wheather it is an Master Card or not
    if first == 5 and second in [1, 2, 3, 4, 5]:
        print("MASTERCARD")
        sys.exit(0)
     
    # check wheather it is an American Express card or not   
    elif first == 3 and (second == 4 or second == 7):
        print("AMEX")
        sys.exit(0)
        
    # check wheather it is an VISA card or not
    elif first == 4:
        print("VISA")
        sys.exit(0)
        
    else:
        print("INVALID")
        
        
main()